0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+  ///
+  /// Full secret:
+  ///      s(y) =  B(node_id,y) = \sum_g_{node_id}(y)
+  fn dealing(&self, churp_id: u8, round: u64);
+
+  /// Construct a reduced share from the switch data points obtained from
+  /// the nodes in the current committee.
+  ///
+  /// Reduced share:
+  ///      r(x) = B(x, node_id)
+  fn share_reduction(&self, churp_id: u8, round: u64);
+
+  /// Randomize reduced share using bivariate shares obtained from
+  /// the nodes in the new committee.
+  ///
+  /// Proactive reduced share:
+  ///      r'(x) = B(x, node_id) + \sum Q(x, node_id)
+  fn proactive_randomization(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the switch data points obtained from 
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId) -> Integer;
+}
+```
+
+Methods:
+
+- `churp/init`
+
+- `churp/verification_matrix`
+
+- `churp/bivariate_share`
+
+- `churp/dealing`
+
+- `churp/share_reduction`
+
+- `churp/proactive_randomization`
+
+- `churp/full_share_distribution`
+
+- `churp/switch_point`
+
+- `churp/key_share`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
+}
+```
+
+### Positive (CHURP)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Dynamic committees.
+
+### Negative (CHURP)
+
+- Slower than KDC.
+
+## Key Derivation Center
+
+Key Derivation Center (KDC) is a secret sharing scheme based on
+the verifiable secret sharing scheme (VSS) where every node possesses
```

Comment:
is verifiability strictly required? why cannot the operators of a paratime/km decide that VSS is not needed, and that they would prefer a simpler algorithm. In fact, that there are already two other proposals here means that a useful artifact for this ADR to depend on would be one describing "key manager-like interfaces" such as in the examples above.

Reply:
> is verifiability strictly required? 

No.

> Why cannot the operators of a paratime/km decide that VSS is not needed, and that they would prefer a simpler algorithm.

Agree, this can and should be optional.

Reply:
Forgot to mention, that clients cannot verify their shares. They can only fetch more shares to find a corrupted one(s).